crates/kitsune_p2p/fetch/src/lib.rs

```diff
@@ -0,0 +1,101 @@
+#![deny(missing_docs)]
+#![deny(unsafe_code)]
+// #![deny(warnings)]
+//! Kitsune P2p Fetch Queue Logic
+
+use kitsune_p2p_types::{dht::region::RegionCoords, KAgent, KOpHash};
+
+mod error;
+mod queue;
+mod respond;
+
+pub use error::*;
+pub use queue::*;
+pub use respond::*;
+
+/// Determine what should be fetched.
+#[derive(
+    Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, serde::Deserialize, serde::Serialize,
+)]
+#[serde(tag = "type", rename_all = "camelCase")]
+pub enum FetchKey {
+    /// Fetch via region.
+    Region {
+        /// The region coordinates to fetch
+        region_coords: RegionCoords,
+    },
+
+    /// Fetch via op hash.
+    Op {
+        /// The hash of the op to fetch.
+        op_hash: KOpHash,
+    },
+}
+
+/// A fetch "unit" that can be de-duplicated.
+#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
+#[serde(tag = "type", rename_all = "camelCase")]
+pub struct FetchRequest {
+    /// Description of what to fetch.
+    pub key: FetchKey,
+
+    /// If specified, the author of the op.
+    /// NOTE: author is additive-only. That is, an op without an author
+    /// is the same as one *with* an author, but should be updated to
+    /// include the author. It is UB to have two FetchKeys with the
+    /// same op_hash, but different authors.
+    pub author: Option<KAgent>,
+
+    /// Optional arguments related to fetching the data.
+    pub options: Option<FetchOptions>,
```

Comment:
If we're not doing gets through this mechanism, should we remove the FetchOptions concept?